homeassistant/components/tuya/fan.py

```diff
@@ -0,0 +1,267 @@
+        self._send_command([{"code": DPCODE_MODE, "value": preset_mode}])
+
+    def set_direction(self, direction: str) -> None:
+        """Set the direction of the fan."""
+        self._send_command([{"code": DPCODE_FAN_DIRECTION, "value": direction}])
+
+    def set_percentage(self, percentage: int) -> None:
+        """Set the speed of the fan, as a percentage."""
+        if self.tuya_device.category == "kj":
+            value_in_range = percentage_to_ordered_list_item(
+                self.air_purifier_speed_range_enum, percentage
+            )
+            self._send_command(
+                [
+                    {
+                        "code": self.dp_code_speed_enum,
+                        "value": value_in_range,
+                    }
+                ]
+            )
+        else:
+            self._send_command([{"code": DPCODE_FAN_SPEED, "value": percentage}])
+
+    def turn_off(self, **kwargs: Any) -> None:
+        """Turn the fan off."""
+        self._send_command([{"code": DPCODE_SWITCH, "value": False}])
+
+    def turn_on(
+        self,
+        speed: str = None,
+        percentage: int = None,
+        preset_mode: str = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn on the fan."""
+        self._send_command([{"code": DPCODE_SWITCH, "value": True}])
+
+    def oscillate(self, oscillating: bool) -> None:
+        """Oscillate the fan."""
+        self._send_command([{"code": DPCODE_SWITCH_HORIZONTAL, "value": oscillating}])
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if fan is on."""
+        return self.tuya_device.status.get(DPCODE_SWITCH, False)
+
+    @property
+    def current_direction(self) -> str:
+        """Return the current direction of the fan."""
+        return (
+            DIRECTION_FORWARD
+            if self.tuya_device.status.get(DPCODE_FAN_DIRECTION)
+            else DIRECTION_REVERSE
+        )
+
+    @property
+    def oscillating(self) -> bool:
+        """Return true if the fan is oscillating."""
+        return self.tuya_device.status.get(DPCODE_SWITCH_HORIZONTAL, False)
+
+    @property
+    def preset_modes(self) -> list:
+        """Return the list of available preset_modes."""
+        if DPCODE_MODE not in self.tuya_device.function:
+            return []
+
+        try:
+            data = json.loads(
+                self.tuya_device.function.get(DPCODE_MODE, {}).values
+            ).get("range")
+            return data
+        except ValueError:
+            _LOGGER.error("Cannot parse the preset modes")
+        return []
+
+    @property
+    def preset_mode(self) -> str:
+        """Return the current preset_mode."""
+        return self.tuya_device.status.get(DPCODE_MODE)
+
+    @property
+    def percentage(self) -> int:
+        """Return the current speed."""
+        if not self.is_on:
+            return 0
+
+        if self.tuya_device.category == "kj":
+            if self.air_purifier_speed_range_len > 1:
```

Comment:
If a check has the same result as the check before it, please combine them with `and`.